controller/userController.js

```diff
@@ -86,7 +92,180 @@ const token = (req, res) => {
+                user.passwordHash = null;
+                user.salt = null;
+                return res.status(HttpStatus.CREATED).json(user);
+            });
+        } else {
+            return res.status(HttpStatus.BAD_REQUEST).json({
+                status: 'Error',
+                message: 'User with provided email already exists!'
+            });
+        }
+    }).catch(err => {
+        const errorMessage = ValidationUtil.buildErrorMessage(err, 'create', 'user');
+        return res.status(HttpStatus.BAD_REQUEST).json({
+            status: 'Error',
+            message: errorMessage
+        });
+    });
+};
+
+const retrieveAll = (req, res) => {
+    User.find({}, { passwordHash: 0, salt: 0 }).then(users => {
+        if (users) {
+            return res.status(HttpStatus.OK).json(users);
+        } else {
+            return res.status(HttpStatus.NOT_FOUND).json({
+                status: 'Error',
+                message: 'Users not found.'
+            });
+        }
+    }).catch(err => {
+        const errorMessage = ValidationUtil.buildErrorMessage(err, 'retrieveAll', 'user');
+        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
+            status: 'Error',
+            message: errorMessage
+        });
+    });
+};
+
+const retrieveById = (req, res) => {
+    const userId = req.params.id;
+    if (userId.length != AppConfig.OBJECT_ID_LEN) {
+        return res.status(HttpStatus.BAD_REQUEST).json({
+            status: 'Error',
+            message: 'Invalid ID'
+        });
+    }
+
+    User.findById(userId, { passwordHash: 0, salt: 0 }).then(user => {
+        if (user) {
+            return res.status(HttpStatus.OK).json(user);
+        } else {
+            return res.status(HttpStatus.NOT_FOUND).json({
+                status: 'Error',
+                message: 'User not found.'
+            });
+        }
+    }).catch(err => {
+        const errorMessage = ValidationUtil.buildErrorMessage(err, 'retrieveById', 'user');
+        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
+            status: 'Error',
+            message: errorMessage
+        });
+    });
+};
+
+const update = (req, res) => {
+    const userId = req.params.id;
+    if (userId.length != AppConfig.OBJECT_ID_LEN) {
+        return res.status(HttpStatus.BAD_REQUEST).json({
+            status: 'Error',
+            message: 'Invalid ID'
+        });
+    }
+    user = {};
+    // Update fields
+    if (req.body.name)
+        user.name = req.body.name;
+    if (req.body.surname)
+        user.surname = req.body.surname;
+    if (req.body.email)
+        user.email = req.body.email;
+    if (req.body.dateOfBirth)
+        user.dateOfBirth = req.body.dateOfBirth;
+    if (req.body.active)
+        user.active = req.body.active;
+    if (req.body.passwordHash) {
+        let retVal = getPasswordHash(req.body.passwordHash);
+        user.salt = retVal.salt;
+        user.passwordHash = retVal.passwordHash;
+    }
```

Comment:
There is no need for this. If values exists and are valid they will be updated. Otherwise they will be not, and an Error will be thrown which is caught in catch. You could simpy use: `const user = req.body`. Also I do not see a check if the update is being done by super-admin/admin of the company or the employee itself. Another thing to consider is which values are allowed to be updated: for example one should not be allowed to change companyID, whilst you should be able to change name or surname. Explore mongoose documentation if some attributes can be prevented from updating.